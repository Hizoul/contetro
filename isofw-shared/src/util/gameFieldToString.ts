import { GameField } from "../gameLogic/gameField"
import { GAMEFIELD_MAX_HEIGHT, GAMEFIELD_MAX_WIDTH } from "../gameLogic/globals"
import { isNumber } from "lodash"

const gameFieldToString = (gameField: GameField, includeHover?: boolean) => {
  let gameFieldString = "\n"
  const pointsInField: any[] = []
  for (let height = 0; height < GAMEFIELD_MAX_HEIGHT; height++) {
    pointsInField[height] = []
    for (let width = 0; width < GAMEFIELD_MAX_WIDTH; width++) {
      pointsInField[height][width] = "o"
    }
  }
  for (const block of gameField.blocks) {
    for (const position of block.getTranslatedBlock()) {
      pointsInField[Number(position.y)][Number(position.x)] = isNumber(block.from) ? block.from : gameField.players.indexOf(block.from)
    }
  }
  const hoverBlock = gameField.getHoverBlock()
  if (includeHover && hoverBlock != null) {
    for (const position of hoverBlock.getTranslatedBlock()) {
      pointsInField[position.y][position.x] = "H"
    }
  }
  for (let width = GAMEFIELD_MAX_HEIGHT - 1; width >= 0; width--) {
    let line = ``
    for (let height = 0; height < GAMEFIELD_MAX_WIDTH; height++) {
      line += pointsInField[width][height]
    }
    gameFieldString += `${line}\n`
  }
  return gameFieldString
}

export default gameFieldToString
