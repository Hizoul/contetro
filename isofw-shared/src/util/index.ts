const mulberry32 = (a: number) => {
  return () => {
    let t = a += 0x6D2B79F5
    t = Math.imul(t ^ t >>> 15, t | 1)
    t ^= t + Math.imul(t ^ t >>> 7, t | 61)
    return ((t ^ t >>> 14) >>> 0) / 4294967296
  }
}

const random = {
  function: mulberry32(Math.floor((Math.random() * Number.MAX_SAFE_INTEGER) + Number.MIN_SAFE_INTEGER)),
  setSeed: (seed: any) => {
    random.function = mulberry32(seed)
  }
}

const randomInRange = (min: number, max: number) => {
  return Math.floor((random.function() * max) + min)
}

const randomColor = () => {
  return `rgba(${randomInRange(0, 255)},${randomInRange(0, 255)},${randomInRange(0, 255)},1)`
}

const range = (from: number, to: number, step: number = 1) => {
  const rangeArray = []
  for (let i = from; i < to; i += step) {
    rangeArray.push(i)
  }
  return rangeArray
}

export {
  randomInRange, range, random, mulberry32, randomColor
}
