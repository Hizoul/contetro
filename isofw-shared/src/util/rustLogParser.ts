import { GameLog } from "isofw-shared/src/gameLogic/gamelog"
import { logType } from "isofw-shared/src/gameLogic/globals"
import { cloneDeep } from "lodash"

const exampleRustLog = `{"log":[{"PayloadPlaced":{"x":0,"y":0,"block":0,"orientation":0,"from":1}},{"PayloadRolled":{"from":0,"block":3}},{"PayloadPlaced":{"x":0,"y":1,"block":3,"orientation":0,"from":0}},{"PayloadRolled":{"from":1,"block":0}},{"PayloadPlaced":{"x":0,"y":3,"block":0,"orientation":0,"from":1}},{"PayloadRolled":{"from":0,"block":1}},{"PayloadPlaced":{"x":0,"y":4,"block":1,"orientation":0,"from":0}},{"PayloadRolled":{"from":1,"block":2}},{"PayloadPlaced":{"x":1,"y":6,"block":2,"orientation":0,"from":1}},{"PayloadRolled":{"from":0,"block":2}},{"PayloadPlaced":{"x":0,"y":8,"block":2,"orientation":1,"from":0}},{"PayloadRolled":{"from":1,"block":0}},{"PayloadPlaced":{"x":1,"y":9,"block":0,"orientation":0,"from":1}},{"PayloadRolled":{"from":0,"block":2}},{"PayloadPlaced":{"x":1,"y":10,"block":2,"orientation":0,"from":0}},{"PayloadRolled":{"from":1,"block":0}},{"PayloadPlaced":{"x":0,"y":12,"block":0,"orientation":0,"from":1}},{"PayloadRolled":{"from":0,"block":4}},{"PayloadPlaced":{"x":0,"y":13,"block":4,"orientation":0,"from":0}},{"PayloadRolled":{"from":1,"block":3}},{"PayloadPlaced":{"x":0,"y":14,"block":3,"orientation":0,"from":1}},{"PayloadRolled":{"from":0,"block":0}},{"PayloadPlaced":{"x":0,"y":16,"block":0,"orientation":0,"from":0}},{"PayloadRolled":{"from":1,"block":1}},{"PayloadPlaced":{"x":0,"y":17,"block":1,"orientation":0,"from":1}},{"PayloadRolled":{"from":0,"block":3}},{"PayloadPlaced":{"x":0,"y":19,"block":3,"orientation":0,"from":0}},{"PayloadRolled":{"from":1,"block":3}},{"PayloadPlaced":{"x":3,"y":17,"block":3,"orientation":0,"from":1}},{"PayloadRolled":{"from":0,"block":4}},{"PayloadPlaced":{"x":3,"y":19,"block":4,"orientation":0,"from":0}},{"PayloadRolled":{"from":1,"block":3}},{"PayloadPlaced":{"x":6,"y":0,"block":3,"orientation":0,"from":1}},{"PayloadRolled":{"from":0,"block":2}},{"PayloadPlaced":{"x":6,"y":2,"block":2,"orientation":1,"from":0}},{"PayloadRolled":{"from":1,"block":1}},{"PayloadPlaced":{"x":6,"y":4,"block":1,"orientation":0,"from":1}},{"PayloadRolled":{"from":0,"block":3}},{"PayloadPlaced":{"x":6,"y":6,"block":3,"orientation":0,"from":0}},{"PayloadRolled":{"from":1,"block":3}},{"PayloadPlaced":{"x":6,"y":8,"block":3,"orientation":0,"from":1}},{"PayloadRolled":{"from":0,"block":3}},{"PayloadPlaced":{"x":6,"y":10,"block":3,"orientation":0,"from":0}},{"PayloadRolled":{"from":1,"block":1}},{"PayloadPlaced":{"x":6,"y":12,"block":1,"orientation":0,"from":1}},{"PayloadRolled":{"from":0,"block":3}},{"PayloadPlaced":{"x":6,"y":14,"block":3,"orientation":0,"from":0}},{"PayloadRolled":{"from":1,"block":0}},{"PayloadPlaced":{"x":6,"y":16,"block":0,"orientation":0,"from":1}},{"PayloadRolled":{"from":0,"block":2}},{"PayloadPlaced":{"x":6,"y":18,"block":2,"orientation":1,"from":0}},{"PayloadRolled":{"from":1,"block":2}},{"PayloadPlaced":{"x":7,"y":20,"block":2,"orientation":2,"from":1}},{"PayloadSkipped":{"from":0,"block":3,"reason":0}},{"PayloadRolled":{"from":1,"block":3}},{"PayloadPlaced":{"x":9,"y":18,"block":3,"orientation":1,"from":1}},{"PayloadRolled":{"from":0,"block":1}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":9,"y":19,"block":1,"orientation":0,"from":0}},{"PayloadSkipped":{"from":1,"block":0,"reason":0}},{"PayloadSkipped":{"from":1,"block":1,"reason":1}},{"PayloadSkipped":{"from":0,"block":3,"reason":0}},{"PayloadSkipped":{"from":1,"block":0,"reason":0}},{"PayloadSkipped":{"from":0,"block":1,"reason":1}}]}`

const scoreExampleRustLog = `{\"log\":[{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":0,\"x\":8,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":0,\"x\":6,\"y\":0}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":0,\"x\":4,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":0,\"x\":7,\"y\":2}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":6,\"y\":2}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":7,\"y\":3}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":6,\"y\":6}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":6,\"y\":10}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":2,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":6,\"y\":14}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":1,\"x\":7,\"y\":5}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":9,\"y\":3}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":3,\"x\":8,\"y\":6}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":10,\"y\":6}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":2,\"y\":2}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":8,\"y\":7}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":2,\"x\":4,\"y\":2}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":1,\"x\":7,\"y\":10}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":0,\"x\":4,\"y\":3}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":2,\"y\":6}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":3,\"x\":5,\"y\":5}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":10}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":3,\"x\":4,\"y\":6}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":3,\"x\":8,\"y\":11}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":1,\"x\":4,\"y\":9}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":3,\"x\":7,\"y\":12}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":1,\"x\":0,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":1,\"x\":4,\"y\":12}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":9,\"y\":12}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":14}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":3,\"x\":0,\"y\":3}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":0,\"y\":4}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":1,\"x\":4,\"y\":14}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":7,\"x\":0,\"y\":5}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":7,\"x\":1,\"y\":8}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":16}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":3,\"y\":4}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":3,\"x\":5,\"y\":15}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":3,\"y\":8}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":18}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":1,\"x\":2,\"y\":12}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":1,\"x\":2,\"y\":15}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":7,\"y\":15}},{\"PayloadSkipped\":{\"from\":1,\"block\":0,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":2,\"x\":3,\"y\":16}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":3,\"x\":0,\"y\":10}},{\"PayloadSkipped\":{\"from\":0,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":2,\"y\":17}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":2,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":4,\"y\":17}},{\"PayloadSkipped\":{\"from\":0,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":2,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":2,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":2,\"y\":19}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":2,\"x\":7,\"y\":18}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadConsidering\":{\"play_index\":0}}]}`


const heuristicLog1 = "{\"log\":[{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":0,\"x\":1,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":1,\"x\":2,\"y\":2}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":0,\"x\":0,\"y\":2}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":2,\"y\":4}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":0,\"x\":5,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":2,\"x\":4,\"y\":0}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadConsidering\":{\"play_index\":0}}]}";

const heuristicLog2 = "{\"log\":[{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":0,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":8,\"y\":0}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":2,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":3,\"y\":2}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":3,\"y\":6}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadConsidering\":{\"play_index\":0}}]}";


const rustPlaced = "PayloadPlaced"
const rustRolled = "PayloadRolled"
const rustSkipped = "PayloadSkipped"
const rustConsidering = "PayloadConsidering"

const payloadConvert = {
  [logType.placed]: rustPlaced,
  [logType.rolled]: rustRolled,
  [logType.skipped]: rustSkipped,
  [logType.considering]: rustConsidering,
  [rustPlaced]: logType.placed,
  [rustSkipped]: logType.skipped,
  [rustRolled]: logType.rolled,
  [rustConsidering]: logType.considering
}

const makeToPlayer = (from: any, players?: string[]) => {
  if (players != null) {
    return players[from]
  }
  return from
}

const rustToJsLog = (log: string, players?: string[]) => {
  const g: GameLog = new GameLog()
  const p: any = JSON.parse(log)
  if (p != null && p.log != null && Array.isArray(p.log)) {
    for (const entry of p.log) {
      if (entry[rustPlaced] != null) {
        const payload = entry[rustPlaced]
        g.blockPlaced(makeToPlayer(payload.from, players), payload.block, payload.orientation, payload.x, payload.y)
      } else if (entry[rustRolled] != null) {
        const payload = entry[rustRolled]
        g.rolledBlock(makeToPlayer(payload.from, players), payload.block)
      } else if (entry[rustSkipped] != null) {
        const payload = entry[rustSkipped]
        g.playerSkipped(makeToPlayer(payload.from, players), payload.block, payload.reason)
      } else if (entry[rustConsidering] != null) {
        const payload = entry[rustConsidering]
        g.log.push({
          type: logType.considering,
          payload: {
            play_index: payload.play_index
          }
        })
      }
    }
  }
  return g
}

const jsToRustLog = (jsLog: GameLog, players: string[]) => {
  const log: any[] = []
  for (const e of cloneDeep(jsLog.log)) {
    if (e.payload != null && e.payload.from) {
      e.payload.from = players.indexOf(e.payload.from)
    }
    if (e.type === logType.placed) {
      log.push({[rustPlaced]: e.payload})
    } else if (e.type === logType.rolled) {
      log.push({[rustRolled]: e.payload})
    } else if (e.type === logType.skipped) {
      log.push({[rustSkipped]: e.payload})
    } else if (e.type === logType.considering) {
      log.push({[rustConsidering]: e.payload})
    }
  }
  return JSON.stringify({log})
}

export {
  exampleRustLog,
  rustToJsLog,
  jsToRustLog,
  scoreExampleRustLog,
  heuristicLog1,
  heuristicLog2
}
