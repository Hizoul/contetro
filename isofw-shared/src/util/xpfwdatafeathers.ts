import {
  FeathersClient
} from "@xpfw/data-feathers"
export default FeathersClient
export {FeathersClient}
