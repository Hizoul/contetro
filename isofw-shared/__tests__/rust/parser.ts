import { exampleRustLog, jsToRustLog, rustToJsLog } from "isofw-shared/src/util/rustLogParser"
import { isEqual } from "lodash"

test("rust conversion back and forth yields same result", () => {
  const g = rustToJsLog(exampleRustLog)
  expect(g).toMatchSnapshot("parsed to js")
  const p: any = [0, 1]
  const rustJson = jsToRustLog(g, p)
  const g2 = rustToJsLog(rustJson)
  expect(isEqual(g2, g)).toBeTruthy()
})
