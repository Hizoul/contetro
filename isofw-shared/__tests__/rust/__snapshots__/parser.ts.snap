// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rust conversion back and forth yields same result: parsed to js 1`] = `
GameLog {
  "log": Array [
    Object {
      "payload": Object {
        "block": 0,
        "from": 1,
        "orientation": 0,
        "x": 0,
        "y": 0,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 3,
        "from": 0,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "block": 3,
        "from": 0,
        "orientation": 0,
        "x": 0,
        "y": 1,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 0,
        "from": 1,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "block": 0,
        "from": 1,
        "orientation": 0,
        "x": 0,
        "y": 3,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 1,
        "from": 0,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "block": 1,
        "from": 0,
        "orientation": 0,
        "x": 0,
        "y": 4,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 2,
        "from": 1,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "block": 2,
        "from": 1,
        "orientation": 0,
        "x": 1,
        "y": 6,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 2,
        "from": 0,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "block": 2,
        "from": 0,
        "orientation": 1,
        "x": 0,
        "y": 8,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 0,
        "from": 1,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "block": 0,
        "from": 1,
        "orientation": 0,
        "x": 1,
        "y": 9,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 2,
        "from": 0,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "block": 2,
        "from": 0,
        "orientation": 0,
        "x": 1,
        "y": 10,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 0,
        "from": 1,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "block": 0,
        "from": 1,
        "orientation": 0,
        "x": 0,
        "y": 12,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 4,
        "from": 0,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "block": 4,
        "from": 0,
        "orientation": 0,
        "x": 0,
        "y": 13,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 3,
        "from": 1,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "block": 3,
        "from": 1,
        "orientation": 0,
        "x": 0,
        "y": 14,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 0,
        "from": 0,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "block": 0,
        "from": 0,
        "orientation": 0,
        "x": 0,
        "y": 16,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 1,
        "from": 1,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "block": 1,
        "from": 1,
        "orientation": 0,
        "x": 0,
        "y": 17,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 3,
        "from": 0,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "block": 3,
        "from": 0,
        "orientation": 0,
        "x": 0,
        "y": 19,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 3,
        "from": 1,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "block": 3,
        "from": 1,
        "orientation": 0,
        "x": 3,
        "y": 17,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 4,
        "from": 0,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "block": 4,
        "from": 0,
        "orientation": 0,
        "x": 3,
        "y": 19,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 3,
        "from": 1,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "block": 3,
        "from": 1,
        "orientation": 0,
        "x": 6,
        "y": 0,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 2,
        "from": 0,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "block": 2,
        "from": 0,
        "orientation": 1,
        "x": 6,
        "y": 2,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 1,
        "from": 1,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "block": 1,
        "from": 1,
        "orientation": 0,
        "x": 6,
        "y": 4,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 3,
        "from": 0,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "block": 3,
        "from": 0,
        "orientation": 0,
        "x": 6,
        "y": 6,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 3,
        "from": 1,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "block": 3,
        "from": 1,
        "orientation": 0,
        "x": 6,
        "y": 8,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 3,
        "from": 0,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "block": 3,
        "from": 0,
        "orientation": 0,
        "x": 6,
        "y": 10,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 1,
        "from": 1,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "block": 1,
        "from": 1,
        "orientation": 0,
        "x": 6,
        "y": 12,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 3,
        "from": 0,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "block": 3,
        "from": 0,
        "orientation": 0,
        "x": 6,
        "y": 14,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 0,
        "from": 1,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "block": 0,
        "from": 1,
        "orientation": 0,
        "x": 6,
        "y": 16,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 2,
        "from": 0,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "block": 2,
        "from": 0,
        "orientation": 1,
        "x": 6,
        "y": 18,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 2,
        "from": 1,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "block": 2,
        "from": 1,
        "orientation": 2,
        "x": 7,
        "y": 20,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 3,
        "from": 0,
        "reason": 0,
      },
      "type": 2,
    },
    Object {
      "payload": Object {
        "block": 3,
        "from": 1,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "block": 3,
        "from": 1,
        "orientation": 1,
        "x": 9,
        "y": 18,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 1,
        "from": 0,
      },
      "type": 0,
    },
    Object {
      "payload": Object {
        "play_index": 0,
      },
      "type": 3,
    },
    Object {
      "payload": Object {
        "block": 1,
        "from": 0,
        "orientation": 0,
        "x": 9,
        "y": 19,
      },
      "type": 1,
    },
    Object {
      "payload": Object {
        "block": 0,
        "from": 1,
        "reason": 0,
      },
      "type": 2,
    },
    Object {
      "payload": Object {
        "block": 1,
        "from": 1,
        "reason": 1,
      },
      "type": 2,
    },
    Object {
      "payload": Object {
        "block": 3,
        "from": 0,
        "reason": 0,
      },
      "type": 2,
    },
    Object {
      "payload": Object {
        "block": 0,
        "from": 1,
        "reason": 0,
      },
      "type": 2,
    },
    Object {
      "payload": Object {
        "block": 1,
        "from": 0,
        "reason": 1,
      },
      "type": 2,
    },
  ],
}
`;
